File: AdventOfCode2022/Day10/day10.js
const utils = require("../utils");
const lines = utils.readInput(__dirname);

let cycles = [];
let x = 1;
let crt = [];
const draw = (crt, spritePosition) => {
  if (crt.length % 40 >= spritePosition - 1 && crt.length % 40 <= spritePosition + 1) {
    crt.push("#");
  } else {
    crt.push(".");
  }
};

for (let i = 0; i < lines.length; i++) {
  const line = lines[i].split(" ");
  if (line[0] === "addx") {
    cycles.push(x);
    draw(crt, x);
    cycles.push(x);
    draw(crt, x);
    x += +line[1];
  } else {
    cycles.push(x);
    draw(crt, x);
  }
}

const p = [20, 60, 100, 140, 180, 220];
let result = p.reduce((r, v) => r + v * cycles[v - 1], 0);
console.log(result);

for (let i = 0; i < crt.length; i += 40) {
  console.log(crt.slice(i, i + 40).join(""));
}

File: AdventOfCode2022/Day10/Day10.hs
import Data.List.Split (chunksOf)

parseAddx line = read (last . words $ line) :: Int

cycleX (x1, _) line = if line == "noop" then (x1, [x1]) else (x1 + parseAddx line, replicate 2 x1)

signalStrength xs i = i * xs !! (i - 1)

draw xs size i = if abs ((xs !! i) - (i `mod` size)) <= 1 then '█' else ' '

main = do
  content <- readFile "input.txt"
  let cycles = concatMap snd $ scanl cycleX (1, []) (lines content)

  let p = [20, 60, 100, 140, 180, 220]
  let part1 = sum $ map (signalStrength cycles) p
  print part1

  let size = 40
  let part2 = map (draw cycles size) [0 .. length cycles - 1]
  putStr $ unlines (chunksOf size part2)
